### DISCUSSION TITLE
How to call Python function in a Tokio request handler

### DISCUSSION THREAD

**Discussion by owenthereal** [PyO3/pyo3#2182] [Questions] [answered]
👋

I'm exploring `pyo3` to see if it's a good fit for what I'm doing. I would like to get a Python script like [this](https://github.com/owenthereal/pyo3func/blob/master/test.py) to work with `pyo3`. I have pushed the example code [here](https://github.com/owenthereal/pyo3func) for the sake of discussion.

I would like that:

* The [`pyo3func.start_function`](https://github.com/owenthereal/pyo3func/blob/e632d7d9565ddae2cb8e8083807b22f03198f2ea/test.py#L7) starts a gRPC server in Rust using [Tonic](https://github.com/hyperium/tonic) (basically an async gRPC server in Tokio)
* When there is a gRPC request to the server, the Tonic handler calls [a Python function](https://github.com/owenthereal/pyo3func/blob/e632d7d9565ddae2cb8e8083807b22f03198f2ea/test.py#L3-L5) and [return the result from the Python function](https://github.com/owenthereal/pyo3func/blob/e632d7d9565ddae2cb8e8083807b22f03198f2ea/src/lib.rs#L50-L55.) as the gRPC response. 

However, I found that [the `Python::with_gil` line hangs forever](https://github.com/owenthereal/pyo3func/blob/e632d7d9565ddae2cb8e8083807b22f03198f2ea/src/lib.rs#L48-L60) in the Tonic/Tokio async handler when there is a request. I'm wondering whether you could help me understand how to make a Python function call in an async Rust handler.

Another issue is that after the Tonic server starts, pressing `ctrl+c` can not stop the running server. I think it's related to that I [block the `pyfunction`](https://github.com/owenthereal/pyo3func/blob/e632d7d9565ddae2cb8e8083807b22f03198f2ea/src/lib.rs#L12-L24). I'm wondering whether there is a better way to start an async Tonic/Tokio server in `pyfunction` that kill signals can reach the Rust server.

I'm fairly new to Pyo3 and I may need some hand-holding 😄 . It would be perfect if you have some example codes that I could take a look at for similar use cases. Thank you in advance!

**Comment by mejrs** [comment, marked as answer]
> I think it's related to that I [block the pyfunction](https://github.com/owenthereal/pyo3func/blob/e632d7d9565ddae2cb8e8083807b22f03198f2ea/src/lib.rs#L12-L24)

Yes, the tokio runtime is multithreaded, and you have this function hold the GIL the entire time, so any other thread will deadlock when they need to acquire it. So you should probably release it before starting the runtime:

```rust
#[pyfunction]
fn start_function(py: Python, f: Py<PyFunction>) -> PyResult<()> {
    let addr = "127.0.0.1:50051".parse().unwrap();
    let func = Function {
        callback: Arc::new(f),
    };
    //              👇 may need `move` here
    py.allow_threads(||{
        let rt = tokio::runtime::Runtime::new().unwrap();
        let serve = Server::builder()
            .add_service(function_server::FunctionServer::new(func))
            .serve(addr);
        rt.block_on(serve).unwrap();
    });

    Ok(())
}
```

Also, the `Arc<Py<...>>` is not needed, you can just use `Py<...>`

**Comment by owenthereal** [comment]
@mejrs:

Thanks for the reply! Yes, adding `py.allow_threads` gets the example to work! And `Arc<Py<...>>` is not needed (only  `Py<...>`). 

> Another issue is that after the Tonic server starts, pressing ctrl+c can not stop the running server

I'm still having trouble sending sigterm to the running server with `ctrl+c` to stop the server from running. Would you have any recommendations or any examples on how to solve it?

**Comment by mejrs** [comment]
You'll probably want to look at https://tokio.rs/tokio/topics/shutdown for that. There is also `Python::check_signals` but given that you probably aren't in Python-land very long I'd just use the tokio handler.
